servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/IdleTimeoutConnectionFilter.java

```diff
@@ -0,0 +1,242 @@
+                timeoutExecutor != null ? timeoutExecutor : contextExecutor(connection.executionContext()));
+    }
+
+    @Override
+    public HttpExecutionStrategy requiredOffloads() {
+        return offloadNone();
+    }
+
+    private static final class ConnectionIdleTimeoutFilterImpl extends StreamingHttpConnectionFilter
+            implements Runnable {
+
+        private static final AtomicIntegerFieldUpdater<ConnectionIdleTimeoutFilterImpl> requestsUpdater =
+                AtomicIntegerFieldUpdater.newUpdater(ConnectionIdleTimeoutFilterImpl.class, "requests");
+        private static final AtomicReferenceFieldUpdater<ConnectionIdleTimeoutFilterImpl, Cancellable>
+                timeoutTaskUpdater = AtomicReferenceFieldUpdater.newUpdater(ConnectionIdleTimeoutFilterImpl.class,
+                Cancellable.class, "timeoutTask");
+
+        private volatile int requests;
+        @Nullable
+        private volatile Cancellable timeoutTask;
+
+        private final long timeoutNs;
+        private final Executor timeoutExecutor;
+
+        // While it may look like "volatile" is not required for this variable for "happens-before" visibility because
+        // we always write to "requests" after updating "lastResponseTime" and read "requests" before reading
+        // "lastResponseTime", non-volatile writes to "long" variables are not atomic and may result in incorrect
+        // reading of the value when 2 threads race.
+        private volatile long lastResponseTime;
+
+        ConnectionIdleTimeoutFilterImpl(final FilterableStreamingHttpConnection connection,
+                                        final long timeoutNs,
+                                        final Executor timeoutExecutor) {
+            super(connection);
+            this.timeoutNs = timeoutNs;
+            this.timeoutExecutor = timeoutExecutor;
+            connection.onClose().whenFinally(this::cancelTask).subscribe();
+            this.lastResponseTime = nanoTime();
+            timeoutTask = this.timeoutExecutor.schedule(this, timeoutNs, NANOSECONDS);
+        }
+
+        private long nanoTime() {
+            return timeoutExecutor.currentTime(NANOSECONDS);
+        }
+
+        private void cancelTask() {
+            final Cancellable oldTask = timeoutTaskUpdater.getAndSet(this, CANCELLED);
+            if (oldTask != null) {
+                oldTask.cancel();
+            }
+        }
+
+        @Override
+        public Single<StreamingHttpResponse> request(final StreamingHttpRequest request) {
+            return defer(() -> {
+                final int inFlightRequests = requestsUpdater.incrementAndGet(this);
+                if (inFlightRequests < 0) {
+                    // To mitigate the tiny risk of going into positive numbers with Integer.MAX_VALUE concurrent
+                    // requests (unrealistic), reset the counter back to MIN_VALUE.
+                    requests = MIN_VALUE;
```

Comment:
if we are protecting against overflow we should do it atomically. Otherwise is likely to cause harm if another thread decrements the value while it is negative we will never get back to 0 as expected.